File: dotnet/Tools/GitHubPodcastEpisodeTool.cs
using System.ComponentModel;
using System.Text;
using System.Xml.Linq;
using GitHub.Copilot;
using Microsoft.Extensions.AI;

namespace CopilotSdkLiveDemo.Tools;

internal static class GitHubPodcastEpisodeTool
{
    private const string FeedUrl = "https://feeds.simplecast.com/ioCY0vfY";

    internal static AIFunction CreateEpisodeTool() =>
        CopilotTool.DefineTool(
            ([Description("Optional GitHub Podcast episode title. Omit for the latest episode.")] string? episodeTitle) =>
                GetAsync(episodeTitle),
            factoryOptions: new AIFunctionFactoryOptions
            {
                Name = "get_github_podcast_episode",
                Description = "Gets a GitHub Podcast episode from the official RSS feed."
            });

    internal static AIFunction CreateLatestEpisodesTool() =>
        CopilotTool.DefineTool(
            () => GetLatestAsync(),
            factoryOptions: new AIFunctionFactoryOptions
            {
                Name = "get_latest_github_podcast_episodes",
                Description = "Gets the ten newest GitHub Podcast episodes from the official RSS feed."
            });

    internal static async Task<IReadOnlyList<EpisodeBrief>> GetLatestAsync()
    {
        var items = await GetItemsAsync();
        return items.Take(10).Select(ToEpisodeBrief).ToList();
    }

    private static async Task<EpisodeBrief> GetAsync(string? episodeTitle)
    {
        var items = await GetItemsAsync();

        var item = string.IsNullOrWhiteSpace(episodeTitle)
            ? items.FirstOrDefault()
            : items.FirstOrDefault(candidate =>
                string.Equals(Value(candidate, "title"), episodeTitle, StringComparison.OrdinalIgnoreCase));

        if (item is null)
        {
            throw new InvalidOperationException(string.IsNullOrWhiteSpace(episodeTitle)
                ? "The GitHub Podcast RSS feed contained no episodes."
                : $"Episode \"{episodeTitle}\" was not found in the GitHub Podcast RSS feed.");
        }

        return ToEpisodeBrief(item);
    }

    private static async Task<List<XElement>> GetItemsAsync()
    {
        using var httpClient = new HttpClient();
        var feed = await httpClient.GetStringAsync(FeedUrl);
        var document = XDocument.Parse(feed);
        return document.Descendants("item").ToList();
    }

    private static EpisodeBrief ToEpisodeBrief(XElement item) =>
        new(
            GetEpisodeNumber(item),
            Value(item, "title"),
            Value(item, "pubDate"),
            item.Elements().FirstOrDefault(element => element.Name.LocalName == "duration")?.Value ?? "Unknown",
            StripHtml(Value(item, "description")),
            Value(item, "link"),
            FeedUrl);

    private static int? GetEpisodeNumber(XElement item)
    {
        var title = Value(item, "title");
        var separator = title.IndexOf(':');
        return separator > 0 && int.TryParse(title[..separator], out var episodeNumber)
            ? episodeNumber
            : null;
    }

    private static string Value(XElement item, string name) => item.Element(name)?.Value.Trim() ?? string.Empty;

    private static string StripHtml(string value)
    {
        var builder = new StringBuilder(value.Length);
        var insideTag = false;

        foreach (var character in value)
        {
            if (character == '<')
            {
                insideTag = true;
            }
            else if (character == '>')
            {
                insideTag = false;
            }
            else if (!insideTag)
            {
                builder.Append(character);
            }
        }

        return System.Net.WebUtility.HtmlDecode(builder.ToString()).Trim();
    }
}

internal sealed record EpisodeBrief(
    int? EpisodeNumber,
    string Title,
    string Published,
    string Duration,
    string Description,
    string EpisodeUrl,
    string SourceFeedUrl);

File: java/src/main/java/demo/GitHubPodcastEpisodeTool.java
package demo;

import com.github.copilot.rpc.ToolDefinition;
import com.github.copilot.tool.Param;

import javax.xml.parsers.DocumentBuilderFactory;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.ArrayList;
import java.util.List;

public final class GitHubPodcastEpisodeTool {
    private static final String FEED_URL = "https://feeds.simplecast.com/ioCY0vfY";

    private GitHubPodcastEpisodeTool() {
    }

    public static ToolDefinition episodeTool() {
        return ToolDefinition.from(
                "get_github_podcast_episode",
                "Gets a GitHub Podcast episode from the official RSS feed.",
                Param.of(String.class, "episodeTitle", "Optional GitHub Podcast episode title. Omit for the latest episode."),
                GitHubPodcastEpisodeTool::getEpisodeJson);
    }

    public static ToolDefinition latestEpisodesTool() {
        return ToolDefinition.from(
                "get_latest_github_podcast_episodes",
                "Gets the ten newest GitHub Podcast episodes from the official RSS feed.",
                GitHubPodcastEpisodeTool::getLatestJson);
    }

    public static List<EpisodeBrief> latest() throws Exception {
        var items = items();
        var episodes = new ArrayList<EpisodeBrief>();
        for (int index = 0; index < Math.min(10, items.size()); index++) {
            episodes.add(toEpisodeBrief(items.get(index)));
        }
        return episodes;
    }

    public static EpisodeBrief pick(List<EpisodeBrief> episodes) throws Exception {
        if (episodes.isEmpty()) {
            throw new IllegalStateException("No GitHub Podcast episodes are available to select.");
        }

        System.out.println("\nChoose a GitHub Podcast episode:");
        for (int index = 0; index < episodes.size(); index++) {
            System.out.println("  " + (index + 1) + ". " + episodes.get(index).title());
        }
        return episodes.get(ModelSelector.readIndex("Episode [1]: ", episodes.size(), 0));
    }

    private static String getLatestJson() {
        try {
            return latest().toString();
        } catch (Exception exception) {
            throw new IllegalStateException("Failed to read The GitHub Podcast RSS feed.", exception);
        }
    }

    private static String getEpisodeJson(String episodeTitle) {
        try {
            for (var item : items()) {
                var episode = toEpisodeBrief(item);
                if (episodeTitle == null || episodeTitle.isBlank() || episodeTitle.equalsIgnoreCase(episode.title())) {
                    return episode.toString();
                }
            }
            throw new IllegalArgumentException("Episode \"" + episodeTitle + "\" was not found in the GitHub Podcast RSS feed.");
        } catch (Exception exception) {
            throw new IllegalStateException("Failed to read The GitHub Podcast RSS feed.", exception);
        }
    }

    private static List<org.w3c.dom.Element> items() throws Exception {
        var response = HttpClient.newHttpClient().send(
                HttpRequest.newBuilder(URI.create(FEED_URL)).build(),
                HttpResponse.BodyHandlers.ofInputStream());
        var document = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(response.body());
        var nodes = document.getElementsByTagName("item");
        var items = new ArrayList<org.w3c.dom.Element>();
        for (int index = 0; index < nodes.getLength(); index++) {
            items.add((org.w3c.dom.Element) nodes.item(index));
        }
        return items;
    }

    private static EpisodeBrief toEpisodeBrief(org.w3c.dom.Element item) {
        var title = value(item, "title");
        return new EpisodeBrief(
                episodeNumber(title),
                title,
                value(item, "pubDate"),
                valueByLocalName(item, "duration", "Unknown"),
                value(item, "description").replaceAll("<[^>]+>", "").trim(),
                value(item, "link"),
                FEED_URL);
    }

    private static Integer episodeNumber(String title) {
        int separator = title.indexOf(':');
        if (separator <= 0) {
            return null;
        }
        try {
            return Integer.parseInt(title.substring(0, separator));
        } catch (NumberFormatException ignored) {
            return null;
        }
    }

    private static String value(org.w3c.dom.Element item, String name) {
        var nodes = item.getElementsByTagName(name);
        return nodes.getLength() == 0 ? "" : nodes.item(0).getTextContent().trim();
    }

    private static String valueByLocalName(org.w3c.dom.Element item, String name, String fallback) {
        var nodes = item.getElementsByTagNameNS("*", name);
        return nodes.getLength() == 0 ? fallback : nodes.item(0).getTextContent().trim();
    }

    public record EpisodeBrief(
            Integer episodeNumber,
            String title,
            String published,
            String duration,
            String description,
            String episodeUrl,
            String sourceFeedUrl) {
    }
}
